codegen/core/src/test/resources/test-files/codegen-annotation/protos/event-sourced-entity/com/example/shoppingcart/shoppingcart_api.proto

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package com.example.shoppingcart; // <1>
+
+import "akkaserverless/annotations.proto"; // <2>
+import "akkaserverless/codegen/annotations.proto"; // <3>
+import "google/api/annotations.proto";
+import "google/protobuf/empty.proto";
+
+option java_outer_classname = "ShoppingCartApi"; // <3>
+
+message AddLineItem { // <4>
+  string cart_id = 1 [(akkaserverless.field).entity_key = true]; // <5>
+  string product_id = 2;
+  string name = 3;
+  int32 quantity = 4;
+}
+
+message RemoveLineItem {
+  string cart_id = 1 [(akkaserverless.field).entity_key = true];
+  string product_id = 2;
+}
+
+message GetShoppingCart {
+  string cart_id = 1 [(akkaserverless.field).entity_key = true];
+}
+
+message LineItem {
+  string product_id = 1;
+  string name = 2;
+  int32 quantity = 3;
+}
+
+message Cart { // <6>
+  repeated LineItem items = 1;
+}
+
+service ShoppingCartService { // <7>
+  option (akkaserverless.codegen) = {
+    event_sourced_entity: {
+        name: "ShoppingCart" 
+        package_name: "com.example.shoppingcart.domain"
```

Comment:
I guess `package` is a reserved word?

Reply:
yes, it's reserved. 

I've been struggling to name this property. 

It's a proto package when we use it to localize the state and events messages, but it's the java package in which we will generate the ShoppingCart entity. 

Ideally, we should have two different fields, but that feels too verbose. We will need to document it well. In any case, we do need to have a codegen section in the docs explaining the codegen.

Reply:
> It's a proto package when we use it to localize the state and events messages, but it's the java package in which we will generate the ShoppingCart entity.

Reply:
> It's a proto package when we use it to localize the state and events messages, but it's the java package in which we will generate the ShoppingCart entity.

That sounds weird. Why should it apply to resolving state and events? Let's try to keep it such that state and event resolution strictly uses proto scope like other types in a proto file as well. Where those live in the Java program should be a different concern.

The question is then if we still need the package annotation at all for generation as well? Couldn't we just remove that setting and generate the entity in the same package that has been declared in the `java_package_name` option for the file? Which kind of usages would that break?

Reply:
The problem is that this is an annotation on the service itself. The service is part of the external API. 

For example, the file above has proto package `com.example.shoppingcart`. The Java class for this service will be generated in `com.example.shoppingcart. ShoppingCartApi.ShoppingCartService` by akka grpc. But this is not really visible to the user, unless they build a client.

And then we have the entity to be generated. It's true that what we generate is a mix of a service and domain. It will have methods that will be derived from the service definition, like `addItem` and `removeItem`. Incoming traffic will be forwarded to those methods. In that sense, they are kind of service methods or external API. 

We could then say, the package for this entity is: `com.example.shoppingcart` and the fully qualified name will be `com.example.shoppingcart.ShoppingCart`.
From there we could say: if you want it in another package, prefix `ShoppingCart` with whatever you want, like `domain.ShoppingCart`. Then it becomes `com.example.shoppingcart.domain.ShoppingCart`.

Same for the state and events. If the user define then on the same package (ie: `com.example.shoppingcart`) all good. If not, they need to prefix them. 

The drawback is that entity, state and events will have a common root package with the service. Not that this is a big deal.

Reply:
@jrudolph, I'm not sure I understand this sentence

> Let's try to keep it such that state and event resolution strictly uses proto scope like other types in a proto file as well. 

Can you clarify? 

For the record, both state and events are pure strings and we can't reference then as proto Messages. It's not that we can import a proto file and reference them.

Reply:
hmm, now I'm in doubt about this: 
> It's not that we can import a proto file and reference them. 

I hope I'm wrong about it. Maybe it's possible to write: 

```scala
option (akkaserverless.codegen) = {
    event_sourced_entity: {
        name: "ShoppingCart" 
        package_name: "com.example.shoppingcart.domain"
        entity_type: "shopping-cart" 
        state: Cart
        events: [
          ItemAdded,
          ItemRemoved ]
    }
  };
```

Reply:
> Can you clarify?

It was about how we resolve state and event, not how to declare them in the proto. It doesn't seem possible to refer to message types in annotation, so they have to be strings. In our resolution code, we can still use the same logic as proto does itself, i.e if the message is defined in the same file, you can refer to it simply by just the name, but if it is defined in another proto of another package, you can also use the fully-qualified (proto) name to reference those (but not using the Java packages).

> The problem is that this is an annotation on the service itself. The service is part of the external API.

Ah, right, so we cannot move the entity definition easily to another file (as before) to change the target package using the per-file option.

Since we now want to bundle service and entity more closely together, do we really need to provide different packages for service and entity after all? Functionally, we need the distinction between `api` and `domain` only to tell messages apart that should be named the same between `api` and `domain` but the API service never turns up in user code anyway.

Reply:
Yeah, I think it's ok to generate the entity in the package defined by the current file. 

What about the following...

```scala
package com.example.shoppingcart; 

service ShoppingCartService {
	  option (akkaserverless.codegen) = {
	      event_sourced_entity: {
	          name: "ShoppingCart" 
	          entity_type: "shopping-cart" 
	          state: "Cart"
	          events: [
	            "ItemAdded",
	            "ItemRemoved" ]
	      }
	    };
...
}
```

Entity generated in `com.example.shoppingcart`. 
State and Events must be define in proto package `com.example.shoppingcart`.

Then, if they are elsewhere...

```scala
package com.example.shoppingcart; 

service ShoppingCartService {
	  option (akkaserverless.codegen) = {
	      event_sourced_entity: {
	          name: "ShoppingCart" 
	          entity_type: "shopping-cart" 
	          state: ".domain.Cart"
	          events: [
	            ".domain.ItemAdded",
	            ".domain.ItemRemoved" ]
	      }
	    };
...
}
```

Meaning the must be in `com.example.shoppingcart.domain`. 

Would that make more sense for you?

Reply:
of course, once could also reference state and events by using a full package, eg `foo.bar.Cart`. In which case, the domain message `Cart` is not sharing any package with the service.

Reply:
So, summing up, with this change, "Entity" and "Service" are not siblings any more in the proto file but one unit. Wouldn't it make sense to try to treat those more as a unit from the point of the user in the first place? So, that there's only a single thing which is (conceptually) "the entity" which provides an API through its (implicit) service.

Couldn't we even remove the `name` then by using a naming convention, e.g. by just appending `Entity` to the service name or similar (even better if we could just use the service name but that probably doesn't work because of the existing akka-grpc code gen)?

```
service ShoppingCart {
  option (akkaserverless.codegen) = {
    event_sourced_entity: {
        entity_type: "shopping-cart" 
        state: "Cart"
        events: [
          "ItemAdded",
          "ItemRemoved" ]
    }
  };

  rpc AddItem (AddLineItem) returns (google.protobuf.Empty) {
    option (google.api.http) = {
      post: "/cart/{cart_id}/items/add"
      body: "*"
    };
  }

  rpc RemoveItem (RemoveLineItem) returns (google.protobuf.Empty) {
    option (google.api.http).post = "/cart/{cart_id}/items/{product_id}/remove";
  }

  rpc GetCart (GetShoppingCart) returns (Cart) {
    option (google.api.http) = {
      get: "/carts/{cart_id}"
      additional_bindings: {
          get: "/carts/{cart_id}/items"
          response_body: "items"
      } };
  }
}
```

Reply:
I think we tried with name conventions before we unsatisfying results. 

We had things like `ShoppingCartService` and `ShoppingCartServiceEntity` which only make things even more convoluted. And the reason is that we needed to sufix it because of conflicts with the akka-grpc generated service.

At the end the user get exposed to the akka-grpc service as well. That happens in the integration tests for instance. So, we should not reuse the name nor suffix it with Entity, IMO. 

Also, if I'm coding a ShoppingCart, I want a class classed ShoppingCart, not ShoppingCartEntity. Entity is a technicality that has nothing to do with the domain I'm modeling. But that's maybe just me. 😄 

In addition to that, I think the new annotations should generate exactly the same code as before. So when updating, I can simply change my protos without touching any of my classes.

Reply:
I'm fine with an optional `name` annotation but we should try remove required options if possible to minimize choice a user has to make before getting started.

> Also, if I'm coding a ShoppingCart, I want a class classed ShoppingCart, not ShoppingCartEntity. Entity is a technicality that has nothing to do with the domain I'm modeling. But that's maybe just me.

I agree, but as we do in other places we can use whatever logic we want to deduce a suitable default name (e.g. if service ends with "Service" just remove it, if it doesn't end with service, add `Entity` etc.).

Reply:
ok, let me rework this a bit. 

I will start by removing the package name. 

We can then experiment with re-using service name (with convention to append entity) and make the name an optional thing. 

Let's see where we get.

Reply:
About the naming convention, I think you are right. We are using the Service name for Actions and Views. 

We will simplify the game if we follow the same convention for entities. Not that the convention is obvious at first, but once the user learn it, they can look to any proto definition and understand what will be generated. 

Btw, we could also add a option `name` field to actions and views to close the circle.

Reply:
Wow, we had it already

```
message ActionDef {
  // Optional name for the view - if not defined, will follow the name of the service
  string name = 1;
}
```

Not sure if we were using this.

Reply:
@jrudolph, the package name is removed as discussed here (see 1568f0fe)

I will work on the change to make the name optional and use the service name instead, but will send it in another PR on top of this as I see it as a change in the current behavior. 

I would like to avoid bloating this PR even further. There is already a lot in here.